### PR TITLE
feat(LiveReload): automatically determines the port to use and DCE

### LINKED ISSUE #1277: [Feature]: Make it so LiveReload automatically detects the correct port
### What is the new or updated feature that you are suggesting?

Currently, the [LiveReload](https://github.com/remix-run/remix/blob/1fd70960e4d88740df5bf407a6ba2cd2b9549459/packages/remix-react/components.tsx#L1273) component accepts a `port` prop that accepts the port for the websocket server. This is a good start, but that port number really shouldn't be something app devs should have to think about. And once we add https://github.com/remix-run/remix/pull/871 (which I think we should) then the port will be non-deterministic.

I suggest we add a mechanism for the LiveReload component to know which port to use (via an environment variable for example).

### Why should this feature be included?

Without this feature, the app developer will have to always specify the port and ensure it's always available which is annoying and kinda makes https://github.com/remix-run/remix/pull/871 less useful.

### PR DESCRIPTION
**Pull request by kentcdodds**
Closes #1277

TL;DR: now devs won't have to ever think about which port the dev server is set to, the `<LiveReload />` component will always use the right one. Combine this with @mcansh's https://github.com/remix-run/remix/pull/871 PR and devs wont even have to worry about configuring ports at all, we'll just choose an available one which will make it much easier for people to run multiple Remix apps at once.

Another thing this PR does is it makes it so devs won't have to put the `process.env.NODE_ENV` guard around their usage of the component anymore. Thanks to the magic of dead code elimination, if we just restructure things a bit we can make it so the production build eliminates our LiveReload component completely in production and devs don't have to think about it one way or another. (I manually tested this is how it works in the gists-app and it worked 😄).

This PR also adds a couple handy tools for our testing setup. Here's the typeahead plugin in action:

[typeahead functionality](https://github.com/jest-community/jest-watch-typeahead)

And here's the project selection in action:

[project selection](https://github.com/jest-community/jest-watch-select-projects).

This makes it a lot easier to slim down what tests you're running in our codebase. I have more ideas beyond this but it's a nice start. Also, we should probably upgrade to jest v27 eventually. It didn't immediately work when I first tried so I punted on that for now.

### CODE REVIEW

**Comment by kentcdodds** [comment]
I'll look into the test failures tomorrow. That happened to me locally when I tried to upgrade to Jest 27. Not sure what's up since this is running Jest 26 still. Must not be related I guess. Would appreciate reviews anyway :)

**Review by mcansh — APPROVED** [review]

**Review by kentcdodds — COMMENTED** [review]
This is ready to go 👍

**Comment by chris-kruining** [comment]
@kentcdodds is there somewhere I can see when this patch will reach stable? I would very much like to use this

**Comment by kentcdodds** [comment]
I believe there's a way to subscribe to releases on GitHub. Once something is merged into dev, it'll be in the next release. We don't have a regular cadence for releases though.

**Review by dpcesx — COMMENTED** [review]
Excellent work!
